### PR TITLE
endpoint to return the courses that a student is enrolled in

### PR DESCRIPTION
**Pull request by jchorl**

### CODE REVIEW

File: app/controllers/courses_controller.rb [hunk 1]
```diff
@@ -24,6 +24,15 @@ def destroy
     head :no_content
   end
 
+  # get visible returns courses that are taught by an admin or enrolled in by a student
+  def get_visible
```
Comment (matthewdu) [inline app/controllers/courses_controller.rb]: I think we should just move this into users controller. Also return this as part of the `show` method.

Comment (matthewdu) [inline app/controllers/courses_controller.rb]: Actually, thinking more about this, this could also work.

Comment (jchorl) [inline app/controllers/courses_controller.rb]: Hmm, we don't want to return everything for the user call. Like we're not also going to return all the assignments and problems. So I suppose we could move this to users, but I'm not totally convinced that it belongs there and I think it makes sense to have a separate call.

File: db/schema.rb [hunk 2]
```diff
@@ -44,6 +44,7 @@
   create_table "problems", force: :cascade do |t|
     t.string "name"
     t.text "description"
+    t.string "cmd"
```
Comment (matthewdu) [inline db/schema.rb]: Why is this here? There isn't a corresponding migration for this.

Comment (jchorl) [inline db/schema.rb]: Need to rebase, one sec

Comment (matthewdu) [inline db/schema.rb]: It's also not on master

Comment (jchorl) [inline db/schema.rb]: fixed

File: spec/requests/assignments_spec.rb [hunk 3]
```diff
@@ -8,6 +8,10 @@
   let(:course) { create(:course) }
   let!(:assignment) { create(:assignment, course_id: course.id) }
 
+  before(:each) do
```
Comment (matthewdu) [inline spec/requests/assignments_spec.rb]: why not create the user in `api_helper.rb` and remove all the before blocks in each spec?

Comment (jchorl) [inline spec/requests/assignments_spec.rb]: we only use before blocks because these specs were all implemented with the assumption that the authenticated user is a student. as we implement tests for teachers in these specs, we'll need to get rid of the before block and decide in each test if it should be a student or teacher that is authenticated.

Comment (matthewdu) [inline spec/requests/assignments_spec.rb]: Ok. Then rename the variable to student?

File: spec/requests/courses_spec.rb [hunk 4]
```diff
@@ -1,19 +1,72 @@
...
+      get "/api/courses/visible"
+      expect(response).to have_http_status(200)
+      parsed = ActiveSupport::JSON.decode(response.body)
+      expect(parsed.size).to eq(1)
+      c = parsed[0]
+      expect(c['id']).to eq(course.id)
+      expect(c['course_code']).to eq(course.course_code)
+      expect(c['title']).to eq(course.title)
+      expect(c['description']).to eq(course.description)
+      expect(c['teacher_id']).to eq(course.teacher_id)
```
Comment (matthewdu) [inline spec/requests/courses_spec.rb]: You can change this to comparing a hash:
I think
```ruby
expect(c).to include(
  id: course.id,
  course_code: course.course_code
  ...
)
```

File: spec/requests/assignments_spec.rb (hunk 3 above)
Comment (jchorl) [inline spec/requests/assignments_spec.rb]: done

File: app/controllers/courses_controller.rb [hunk 5]
```diff
@@ -24,6 +24,15 @@ def destroy
...
+    if current_user.admin
```
Comment (rwongone) [inline app/controllers/courses_controller.rb]: We could make this one line:

`render status: :ok, json: current_user.taught_courses if current_user.admin else current_user.enrolled_courses`

File: spec/factories/courses.rb [hunk 6]
```diff
@@ -1,6 +1,6 @@
 FactoryGirl.define do
   factory :course do
-    association :teacher, factory: [:user, :teacher]
+    association :teacher, factory: [:user, :teacher], email: "somerandomemail@r.com", google_id: "sdkfjdsl2342342"
```
Comment (rwongone) [inline spec/factories/courses.rb]: Just a personal preference, but might be good to make the string itself indicate that it's a google_id: "google_id123" for instance.

File: spec/requests/assignments_spec.rb (hunk 3 above)
Comment (rwongone) [inline spec/requests/assignments_spec.rb]: Good call there, I like it :)

File: spec/requests/courses_spec.rb [hunk 7]
```diff
@@ -1,19 +1,73 @@
...
+require 'helpers/api_helper'
 
 RSpec.describe "Courses", type: :request do
   include_context "with authenticated requests"
-  let(:user) { create(:user) }
+  let(:student) { create(:user) }
+  let(:teacher) { create(:user, :teacher) }
   let!(:course) { create(:course) }
 
-  before do
```
Comment (rwongone) [inline spec/requests/courses_spec.rb]: In this set of specs, we can
`before do
  authenticate(user)
end
`
and in each test case, `let(:user) { create(:user) }`, or `let(:user) { create(:user, :teacher) }`. The spec will used the "user" specified in the scope most specific to the spec.

We may also consider making different factories entirely for student and teacher, instead of having teacher be a trait.

File: spec/requests/courses_spec.rb [hunk 8]
```diff
@@ -1,19 +1,73 @@
...
       get "/api/courses/#{course.id}"
       expect(response).to have_http_status(200)
-      expect(response.body).to eq(course.to_json)
+      parsed = ActiveSupport::JSON.decode(response.body)
+      validate_course(parsed, course)
+    end
+  end
+
+  describe "GET /api/courses/visible" do
+    it "returns the Courses that the student is enrolled in as JSON" do
```
Comment (rwongone) [inline spec/requests/courses_spec.rb]: Let's make this into two separate cases: one where the student is enrolled in something, and one where the student is not. Go for clarity of the test over DRY.

File: spec/requests/courses_spec.rb [hunk 9]
```diff
@@ -1,19 +1,73 @@
...
+
+      # ensure courses are being returned
+      get "/api/courses/visible"
+      expect(response).to have_http_status(200)
+      parsed = ActiveSupport::JSON.decode(response.body)
+      expect(parsed.size).to eq(1)
+      validate_course(parsed[0], course)
+    end
+
+    it "returns the Courses that the teacher teaches as JSON" do
```
Comment (rwongone) [inline spec/requests/courses_spec.rb]: Same idea goes here: let's make this two specs.

File: spec/requests/courses_spec.rb (hunk 7 above)
Comment (rwongone) [inline spec/requests/courses_spec.rb]: Alternatively, in the before block, we could authenticate a student and a teacher and just work with those in the `describe`s.

File: app/controllers/courses_controller.rb (hunk 5 above)
Comment (matthewdu) [inline app/controllers/courses_controller.rb]: Let's keep it two lines for readability

Comment (jchorl) [inline app/controllers/courses_controller.rb]: agreed. we need to think hard about how much we separate student and teacher logic. we absolutely cannot afford to mix them up.

File: spec/requests/courses_spec.rb (hunk 7 above)
Comment (jchorl) [inline spec/requests/courses_spec.rb]: I'm trying to stay away from using variables called `user`. I think it should be very clear when a var is a student or a teacher. I dont think we can authenticate both a student and a teacher at the same time because the mock must return one value or the other.

Comment (rwongone) [inline spec/requests/courses_spec.rb]: Ah, yes that's true. In that case, this should suffice. Or, you can split the specs into two contexts, one where a student is authenticated, and one where a teacher is authenticated.

File: app/controllers/courses_controller.rb (hunk 5 above)
Comment (rwongone) [inline app/controllers/courses_controller.rb]: Ok, fine by me.
